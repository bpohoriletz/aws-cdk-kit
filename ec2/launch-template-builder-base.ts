/* eslint-disable @typescript-eslint/no-unused-expressions */
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as iam from 'aws-cdk-lib/aws-iam';
import { LaunchTemplateProduct } from './products/launch-template';

export default abstract class LaunchTemplateBuilderBase {
  protected launchTemplateProps: LaunchTemplateProduct;

  constructor() {
    this.launchTemplateProps = new LaunchTemplateProduct();
  }

  setInstanceProfile(profile?: iam.IInstanceProfile): this {
    profile && (this.launchTemplateProps.instanceProfile = profile);

    return this;
  }

  setInstanceType(instanceType?: ec2.InstanceType): this {
    instanceType && (this.launchTemplateProps.instanceType = instanceType);

    return this;
  }

  setName(name?: string): this {
    name && (this.launchTemplateProps.launchTemplateName = name);

    return this;
  }

  setSecurityGroup(group?: ec2.SecurityGroup): this {
    group && (this.launchTemplateProps.securityGroup = group);

    return this;
  }

  getResult(): ec2.LaunchTemplateProps {
    return this.launchTemplateProps;
  }
}
