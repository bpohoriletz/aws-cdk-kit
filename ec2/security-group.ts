import { Stack } from "aws-cdk-lib";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as con from "../naming/resources"

// Create Security Group for web instances
export function createWebSg(resourceNamePrefix: string[], vpc: ec2.IVpc, stack: Stack): ec2.SecurityGroup {
  let resourceName: string = con.ec2SecurityGroupName(resourceNamePrefix.slice(0,3), "web")
  const webSecurityGroup = new ec2.SecurityGroup(stack, resourceName, {
    vpc: vpc,
    description: "Security Group for the Web instances",
    securityGroupName: resourceName,
    allowAllOutbound: true
  })

  return webSecurityGroup
}

export function createPublicWebSg(resourceNamePrefix: string[], vpc: ec2.IVpc, stack: Stack): ec2.SecurityGroup {
  let resourceName: string = con.ec2SecurityGroupName(resourceNamePrefix.slice(0,3), "web")
  const publicWebSecurityGroup = new ec2.SecurityGroup(stack, resourceName, {
    vpc: vpc,
    description: "Security Group for public EC2 Web instances",
    securityGroupName: resourceName,
    allowAllOutbound: true
  })
  publicWebSecurityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(80), "Allow incoming traffic over port 80");
  publicWebSecurityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(443), "Allow incoming traffic over port 443");

  return publicWebSecurityGroup
}
