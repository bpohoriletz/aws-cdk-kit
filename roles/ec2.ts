import { Role, RoleProps, ServicePrincipal, ManagedPolicy, Policy } from 'aws-cdk-lib/aws-iam';
import { Stack } from 'aws-cdk-lib';
import PolicyStatementDirector from '../directors/policy-statement';
import S3PolicyStatementBuilder from '../iam/policy-statement-builders/s3';
import LogsPolicyStatementBuilder from '../iam/policy-statement-builders/logs';
import ElasticBPolicyStatementBuilder from '../iam/policy-statement-builders/elasticb';
import * as con from '../utils/naming';

export function createEc2Role(resourceNamePrefix: string[], bucketArns: string[], stack: Stack): Role {
  const roleProps: RoleProps = {
    assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
  };
  const ec2Role = new Role(stack, con.iamInstanceRoleName(resourceNamePrefix), roleProps);
  ec2Role.addManagedPolicy(
    ManagedPolicy.fromManagedPolicyArn(stack, stack.stackId, 'arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore')
  );

  const resources = bucketArns.flatMap((arn) => [arn, `${arn}/resources/*`]);
  new Policy(stack, 'allowEc2UploadLogsToEbRegionalBucket', {
    statements: [new PolicyStatementDirector(S3PolicyStatementBuilder).constructS3ReadWritePolicyStatement(resources)],
  }).attachToRole(ec2Role);

  new Policy(stack, 'allowEc2PutStatistics', {
    statements: [new PolicyStatementDirector(ElasticBPolicyStatementBuilder).constructEbEC2StatsPolicyStatement()],
  }).attachToRole(ec2Role);

  new Policy(stack, 'allowEc2UploadLogs', {
    statements: [new PolicyStatementDirector(LogsPolicyStatementBuilder).constructLogsUploadPolicyStatement()],
  }).attachToRole(ec2Role);

  return ec2Role;
}
