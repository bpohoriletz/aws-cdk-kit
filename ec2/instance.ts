import { Stack } from "aws-cdk-lib";
import { InstanceProfile } from "aws-cdk-lib/aws-iam";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as iam from "aws-cdk-lib/aws-iam";
import { createPublicWebSg }  from "../ec2/security-group";
import * as con from "../naming/resources";

export function createPublicEc2(
  role: iam.Role,
  resourceNamePrefix: string[],
  stack: Stack,
  vpc: ec2.IVpc,
  instanceType: ec2.InstanceType = new ec2.InstanceType('t2.micro')) : [ec2.Instance, ec2.SecurityGroup] {

    const ec2Sg: ec2.SecurityGroup = createPublicWebSg(resourceNamePrefix, vpc, stack)

    const ec2Instance: ec2.Instance = new ec2.Instance(stack, con.ec2Name(resourceNamePrefix, 'public'), {
      associatePublicIpAddress: true,
      instanceType: instanceType,
      machineImage: ec2.MachineImage.latestAmazonLinux2023(),
      role: role,
      securityGroup: ec2Sg,
      vpc: vpc,
      vpcSubnets: { subnetType: ec2.SubnetType.PUBLIC },
    });

    return [ec2Instance, ec2Sg]
  }
