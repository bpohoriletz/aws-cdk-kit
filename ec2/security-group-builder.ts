import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { ISecurityGroupBuilder, SecurityGroupProduct } from '../products/security-group';

export default class SecurityGroupBuilder implements ISecurityGroupBuilder {
  protected securityGroupProps: SecurityGroupProduct;

  constructor() {
    this.securityGroupProps = new SecurityGroupProduct();
  }

  customizeVpc(vpc: ec2.IVpc): this {
    this.securityGroupProps.setVpcs(vpc);

    return this;
  }

  customizeName(name: string): this {
    this.securityGroupProps.setSecurityGroupName(name);

    return this;
  }

  customizeDescription(description: string): this {
    this.securityGroupProps.setDescription(description);

    return this;
  }

  getResult(): SecurityGroupProduct {
    return this.securityGroupProps;
  }
}
