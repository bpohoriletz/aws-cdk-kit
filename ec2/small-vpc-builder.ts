import * as ec2 from "aws-cdk-lib/aws-ec2";
import { IVpcBuilder, VpcProduct } from "../products/vpc";
import * as conf from "../../config/ec2/vpc";

export default class SmallVpcBuilder implements IVpcBuilder {
  private resourceName: string;
  private vpcProps: VpcProduct;

  constructor(resourceName: string) {
    this.resourceName = resourceName;
    this.vpcProps = new VpcProduct();
  }

  customizeCidr(): this {
    this.vpcProps.setCidr(conf.vpcCidrConfig(this.resourceName));

    return this;
  }
  customizeAzs(): this {
    this.vpcProps.setMaxAzs(1);

    return this;
  }
  customizeNatProvider(): this {
    const provider = ec2.NatProvider.instanceV2({
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.T2,
        ec2.InstanceSize.NANO
      ),
    });
    this.vpcProps.setNatProvider(provider);

    return this;
  }
  customizeNatSubnets(): this {
    this.vpcProps.setNatSubnets({ subnetType: ec2.SubnetType.PUBLIC });

    return this;
  }
  customizeSubnetConfiguration(): this {
    const config = conf.vpcSubnetConfig(this.resourceName);
    this.vpcProps.setSubnetConfiguration(config)

    return this;
  }

  getResult(): VpcProduct {
    return this.vpcProps;
  };
}
