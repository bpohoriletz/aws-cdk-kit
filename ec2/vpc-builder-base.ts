import { IVpcBuilder, VpcProduct } from '../products/vpc';
import * as conf from '../../config/ec2/vpc';

export default class VpcBuilder implements IVpcBuilder {
  protected vpcProps: VpcProduct;

  constructor() {
    this.vpcProps = new VpcProduct();
  }

  setCidr(name: string): this {
    this.vpcProps.ipAddresses = conf.vpcCidrConfig(name);

    return this;
  }

  setSubnetConfiguration(name: string): this {
    this.vpcProps.subnetConfiguration = conf.vpcSubnetConfig(name);

    return this;
  }

  getResult(): VpcProduct {
    return this.vpcProps;
  }
}
