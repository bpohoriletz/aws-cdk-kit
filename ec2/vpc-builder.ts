import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { IVpcBuilder, VpcProduct } from '../products/vpc';
import * as conf from '../../config/ec2/vpc';

export default class VpcBuilder implements IVpcBuilder {
  protected vpcProps: VpcProduct;

  constructor() {
    this.vpcProps = new VpcProduct();
  }

  customizeCidr(name: string): this {
    this.vpcProps.setCidr(conf.vpcCidrConfig(name));

    return this;
  }
  customizeAzs(): this {
    this.vpcProps.setMaxAzs(2);

    return this;
  }
  customizeNatProvider(): this {
    const provider = ec2.NatProvider.instanceV2({
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.T2, ec2.InstanceSize.NANO),
    });
    this.vpcProps.setNatProvider(provider);

    return this;
  }
  customizeNatSubnets(): this {
    this.vpcProps.setNatSubnets({ subnetType: ec2.SubnetType.PUBLIC });

    return this;
  }
  customizeSubnetConfiguration(name: string): this {
    const config = conf.vpcSubnetConfig(name);
    this.vpcProps.setSubnetConfiguration(config);

    return this;
  }

  getResult(): VpcProduct {
    return this.vpcProps;
  }
}
