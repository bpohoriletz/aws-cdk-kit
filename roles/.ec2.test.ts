import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { arn } from "../test/stubs";

import * as func from "./ec2";

describe("IAM .createEc2Role()", () => {
  let stack: cdk.Stack;

  beforeEach(() => {
    stack = new cdk.Stack();
  });

  test("matches snapshot", () => {
    func.createEc2Role(["pre", "fix"], [arn()], stack);

    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});
