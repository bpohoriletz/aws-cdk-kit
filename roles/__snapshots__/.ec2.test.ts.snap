// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IAM .createEc2Role() matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "PreFixEc2InstRoleF8B48E19": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "allowEc2PutStatisticsD9C1A435": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "elasticbeanstalk:PutInstanceStatistics",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "allowEc2PutStatisticsD9C1A435",
        "Roles": [
          {
            "Ref": "PreFixEc2InstRoleF8B48E19",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "allowEc2UploadLogsA49BD994": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "allowEc2UploadLogsA49BD994",
        "Roles": [
          {
            "Ref": "PreFixEc2InstRoleF8B48E19",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "allowEc2UploadLogsToEbRegionalBucket40230F17": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutObject",
                "s3:ListBucket",
                "s3:ListBucketVersions",
                "s3:GetObject",
                "s3:GetObjectVersion",
              ],
              "Effect": "Allow",
              "Resource": [
                "arn:partition:service:region:account-id:resource",
                "arn:partition:service:region:account-id:resource/resources/*",
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "allowEc2UploadLogsToEbRegionalBucket40230F17",
        "Roles": [
          {
            "Ref": "PreFixEc2InstRoleF8B48E19",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
